# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/components/extensions/parent/ext-cookies.js
# Commit: 36aa6faea947
# Full Hash: 36aa6faea9477a13b6d94e396596697e56675018
# Author: Paul Zuehlcke <pbz@mozilla.com>
# Date: 2023-09-19 03:53:26
# Regressor Bug: 1839918
# File Overlap Count: 6
# Description:
#   Bug 1839918 - Refactor "cookie-changed" notifications to use nsICookieNotification. r=extension-reviewers,cookie-reviewers,sessionstore-reviewers,bvandersloot,edgul,robwu
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D185300
# ==============================================================================

diff -r 905592ffa91d -r 36aa6faea947 toolkit/components/extensions/parent/ext-cookies.js
--- a/toolkit/components/extensions/parent/ext-cookies.js	Mon Sep 18 11:25:24 2023 +0000
+++ b/toolkit/components/extensions/parent/ext-cookies.js	Mon Sep 18 11:55:08 2023 +0000
@@ -454,7 +454,7 @@
 this.cookies = class extends ExtensionAPIPersistent {
   PERSISTENT_EVENTS = {
     onChanged({ fire }) {
-      let observer = (subject, topic, data) => {
+      let observer = (subject, topic) => {
         let notify = (removed, cookie, cause) => {
           cookie.QueryInterface(Ci.nsICookie);
 
@@ -470,22 +470,34 @@
           }
         };
 
+        let notification = subject.QueryInterface(Ci.nsICookieNotification);
+        let { cookie } = notification;
+
+        let {
+          COOKIE_DELETED,
+          COOKIE_ADDED,
+          COOKIE_CHANGED,
+          COOKIES_BATCH_DELETED,
+        } = Ci.nsICookieNotification;
+
         // We do our best effort here to map the incompatible states.
-        switch (data) {
-          case "deleted":
-            notify(true, subject, "explicit");
-            break;
-          case "added":
-            notify(false, subject, "explicit");
+        switch (notification.action) {
+          case COOKIE_DELETED:
+            notify(true, cookie, "explicit");
             break;
-          case "changed":
-            notify(true, subject, "overwrite");
-            notify(false, subject, "explicit");
+          case COOKIE_ADDED:
+            notify(false, cookie, "explicit");
+            break;
+          case COOKIE_CHANGED:
+            notify(true, cookie, "overwrite");
+            notify(false, cookie, "explicit");
             break;
-          case "batch-deleted":
-            subject.QueryInterface(Ci.nsIArray);
-            for (let i = 0; i < subject.length; i++) {
-              let cookie = subject.queryElementAt(i, Ci.nsICookie);
+          case COOKIES_BATCH_DELETED:
+            let cookieArray = notification.batchDeletedCookies.QueryInterface(
+              Ci.nsIArray
+            );
+            for (let i = 0; i < cookieArray.length; i++) {
+              let cookie = cookieArray.queryElementAt(i, Ci.nsICookie);
               if (!cookie.isSession && cookie.expiry * 1000 <= Date.now()) {
                 notify(true, cookie, "expired");
               } else {
